File: binary-search-tree/binary-search-tree.py
class TreeNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        if self.root is None:
            self.root = TreeNode(value)
        else:
            self._insert(self.root, value)

    def _insert(self, root, value):
        if value < root.data:
            if root.left is None:
                root.left = TreeNode(value)
            else:
                self._insert(root.left, value)
        else:
            if root.right is None:
                root.right = TreeNode(value)
            else:
                self._insert(root.right, value)

    def inorder(self):
        self._inorder(self.root)

    def _inorder(self, root):
        if root:
            self._inorder(root.left)
            print(root.data, end=" ")
            self._inorder(root.right)

    def preorder(self):
        self._preorder(self.root)

    def _preorder(self, root):
        if root:
            print(root.data, end=" ")
            self._preorder(root.left)
            self._preorder(root.right)

    def postorder(self):
        self._postorder(self.root)

    def _postorder(self, root):
        if root:
            self._postorder(root.left)
            self._postorder(root.right)
            print(root.data, end=" ")

    def search(self, key):
        return self._search(self.root, key)

    def _search(self, root, key):
        if root is None or root.data == key:
            return root
        if key < root.data:
            return self._search(root.left, key)
        return self._search(root.right, key)

    def delete(self, key):
        self.root = self._delete(self.root, key)

    def _delete(self, root, key):
        if root is None:
            return root
        if key < root.data:
            root.left = self._delete(root.left, key)
        elif key > root.data:
            root.right = self._delete(root.right, key)
        else:
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            root.data = self._find_min(root.right)
            root.right = self._delete(root.right, root.data)
        return root

    def _find_min(self, root):
        while root.left:
            root = root.left
        return root.data

if __name__ == "__main__":
    tree = BinaryTree()
    while True:
        print("1. Insert\n2. Delete\n3. Search\n4. Inorder\n5. Preorder\n6. Postorder\n7. Exit")
        choice = int(input("Enter your choice: "))
        if choice == 1:
            ele = int(input("Enter element to insert: "))
            tree.insert(ele)
        elif choice == 2:
            ele = int(input("Enter element to delete: "))
            tree.delete(ele)
        elif choice == 3:
            ele = int(input("Enter element to search: "))
            result = tree.search(ele)
            if result:
                print(f"{ele} found in the tree.")
            else:
                print(f"{ele} not found in the tree.")
        elif choice == 4:
            print("Inorder Traversal: ", end="")
            tree.inorder()
            print()
        elif choice == 5:
            print("Preorder Traversal: ", end="")
            tree.preorder()
            print()
        elif choice == 6:
            print("Postorder Traversal: ", end="")
            tree.postorder()
            print()
        elif choice == 7:
            break
        else:
            print("Invalid choice!")

File: binary-search-tree/binary-search-tree.java
import java.util.Scanner;

class TreeNode {
    int data;
    TreeNode left, right;

    public TreeNode(int data) {
        this.data = data;
        this.left = this.right = null;
    }
}

class BinaryTree {
    private TreeNode root;

    public void insert(int value) {
        root = insertRec(root, value);
    }

    private TreeNode insertRec(TreeNode root, int value) {
        if (root == null) {
            root = new TreeNode(value);
            return root;
        }
        if (value < root.data)
            root.left = insertRec(root.left, value);
        else if (value > root.data)
            root.right = insertRec(root.right, value);
        return root;
    }

    public void inorder() {
        inorderRec(root);
    }

    private void inorderRec(TreeNode root) {
        if (root != null) {
            inorderRec(root.left);
            System.out.print(root.data + " ");
            inorderRec(root.right);
        }
    }

    public void preorder() {
        preorderRec(root);
    }

    private void preorderRec(TreeNode root) {
        if (root != null) {
            System.out.print(root.data + " ");
            preorderRec(root.left);
            preorderRec(root.right);
        }
    }

    public void postorder() {
        postorderRec(root);
    }

    private void postorderRec(TreeNode root) {
        if (root != null) {
            postorderRec(root.left);
            postorderRec(root.right);
            System.out.print(root.data + " ");
        }
    }

    public TreeNode search(int key) {
        return searchRec(root, key);
    }

    private TreeNode searchRec(TreeNode root, int key) {
        if (root == null || root.data == key) return root;
        if (key < root.data)
            return searchRec(root.left, key);
        return searchRec(root.right, key);
    }

    public void delete(int key) {
        root = deleteRec(root, key);
    }

    private TreeNode deleteRec(TreeNode root, int key) {
        if (root == null) return root;
        if (key < root.data)
            root.left = deleteRec(root.left, key);
        else if (key > root.data)
            root.right = deleteRec(root.right, key);
        else {
            if (root.left == null) return root.right;
            else if (root.right == null) return root.left;
            root.data = findMin(root.right);
            root.right = deleteRec(root.right, root.data);
        }
        return root;
    }

    private int findMin(TreeNode root) {
        while (root.left != null)
            root = root.left;
        return root.data;
    }
}

public class BST {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        BinaryTree tree = new BinaryTree();

        while (true) {
            System.out.println("1. Insert\n2. Delete\n3. Search\n4. Inorder\n5. Preorder\n6. Postorder\n7. Exit");
            System.out.print("Enter your choice: ");
            int choice = sc.nextInt();

            switch (choice) {
                case 1:
                    System.out.print("Enter element to insert: ");
                    int ele = sc.nextInt();
                    tree.insert(ele);
                    break;
                case 2:
                    System.out.print("Enter element to delete: ");
                    ele = sc.nextInt();
                    tree.delete(ele);
                    break;
                case 3:
                    System.out.print("Enter element to search: ");
                    ele = sc.nextInt();
                    TreeNode result = tree.search(ele);
                    if (result != null)
                        System.out.println(ele + " found in the tree.");
                    else
                        System.out.println(ele + " not found in the tree.");
                    break;
                case 4:
                    System.out.print("Inorder Traversal: ");
                    tree.inorder();
                    System.out.println();
                    break;
                case 5:
                    System.out.print("Preorder Traversal: ");
                    tree.preorder();
                    System.out.println();
                    break;
                case 6:
                    System.out.print("Postorder Traversal: ");
                    tree.postorder();
                    System.out.println();
                    break;
                case 7:
                    System.exit(0);
            }
        }
    }
}
